### PR TITLE
Reduce the number of retain cycles between host and guest

### PR DESCRIPTION
**Pull request by swankjesse**
---

- [x] `CHANGELOG.md`'s "Unreleased" section has been updated, if applicable.

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -12,6 +12,7 @@ Changed:
 - APIs accepting a `FileSystem` and `Path` now have the `FileSystem` coming before the `Path` in the parameter list. Compatibility functions are retained for this version, but will be removed in the next version.
 
 Fixed:
+- Fix memory leaks caused by reference cycles on iOS. We got into trouble mixing garbage-collected Kotlin objects with reference-counted Swift objects.
```
Comment (JakeWharton) [inline CHANGELOG.md]: Can you do a quick rebase? I released these notes a few hours ago and there's a new "Unreleased" section now.

**Review by JakeWharton — APPROVED** [review]
Nice

File: CHANGELOG.md (hunk 1 above)
Comment (swankjesse) [reply CHANGELOG.md]: Oooh yep. Done.

**Comment by swankjesse** [comment]
Ooooh darn.
```
e: file:///Users/runner/work/redwood/redwood/test-app/schema/protocol-host/build/generated/redwood/sources/com/example/redwood/testapp/protocol/host/redwoodlazylayout/ProtocolRefreshableLazyList.kt:112:17 Implementing function interface is prohibited in JavaScript
```

File: redwood-tooling-codegen/src/main/kotlin/app/cash/redwood/tooling/codegen/protocolHostGeneration.kt [hunk 2]
```diff
@@ -371,30 +377,25 @@ internal fun generateProtocolNode(
...
-                          "json.encodeToJsonElement(serializer_%L, arg%L)",
-                          serializerId,
-                          index,
-                        )
+                        arguments += CodeBlock.of("serializer_%L", serializerId)
                       }
                       if (trait.parameterTypes.isEmpty()) {
-                        beginControlFlow("{")
+                        addStatement(
+                          "%L(json, change.id, eventSink)::invoke",
```
Comment (JakeWharton) [inline redwood-tooling-codegen/src/main/kotlin/app/cash/redwood/tooling/codegen/protocolHostGeneration.kt:389]: Uuuuuggghhh I hate it. So wasteful. I wonder if ES6 or K2 or both will remove that implementation limitation for future optimization.

Comment (swankjesse) [reply redwood-tooling-codegen/src/main/kotlin/app/cash/redwood/tooling/codegen/protocolHostGeneration.kt:389]: Same.

Comment (JakeWharton) [reply redwood-tooling-codegen/src/main/kotlin/app/cash/redwood/tooling/codegen/protocolHostGeneration.kt:389]: https://youtrack.jetbrains.com/issue/KT-29580/JS-allow-to-inherit-from-Function-interfaces-in-JS-IR-BE

**Pull request merged**